crates/circuit/src/bit.rs

```diff
@@ -0,0 +1,1078 @@
+                    uid,
+                    subclass: QubitSubclass::ANCILLA,
+                })))
+                .add_subclass(PyAncillaQubit),
+        )?
+        .into_any())
+    }
+    #[staticmethod]
+    fn _from_owned(
+        py: Python,
+        reg_name: String,
+        reg_size: u32,
+        index: u32,
+    ) -> PyResult<Bound<PyAny>> {
+        // This doesn't feel like the most efficient way - in a big list of owned qubits,
+        // we'll pickle the register information many times - but good enough for now.
+        let register = Arc::new(OwningRegisterInfo {
+            name: reg_name,
+            size: reg_size,
+            subclass: QubitSubclass::ANCILLA,
+        });
+        Ok(Bound::new(
+            py,
+            PyClassInitializer::from(PyBit)
+                .add_subclass(PyQubit(ShareableQubit(BitInfo::Owned { register, index })))
+                .add_subclass(PyAncillaQubit),
+        )?
+        .into_any())
+    }
+}
+impl<'py> IntoPyObject<'py> for ShareableQubit {
+    type Target = PyAny;
+    type Output = Bound<'py, PyAny>;
+    type Error = PyErr;
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        if self.is_ancilla() {
+            Ok(Bound::new(
+                py,
+                PyClassInitializer::from(PyBit)
+                    .add_subclass(PyQubit(self))
+                    .add_subclass(PyAncillaQubit),
+            )?
+            .into_any())
+        } else {
+            Ok(Bound::new(py, (PyQubit(self), PyBit))?.into_any())
+        }
+    }
+}
+
+impl QuantumRegister {
+    /// Create a new ancilla register that owns its qubits.
+    pub fn new_ancilla_owning(name: String, size: u32) -> Self {
+        Self(Arc::new(RegisterInfo::Owning(Arc::new(
+            OwningRegisterInfo {
+                name,
+                size,
+                subclass: QubitSubclass::ANCILLA,
+            },
+        ))))
+    }
+
+    /// Create a new ancilla register that aliases other bits.
+    ///
+    /// Returns `None` if not all the bits are ancillas.
+    pub fn new_ancilla_alias(name: String, bits: Vec<ShareableQubit>) -> Option<Self> {
+        bits.iter().all(|bit| bit.is_ancilla()).then(|| {
+            Self(Arc::new(RegisterInfo::Alias {
+                name,
+                bits,
+                subclass: QubitSubclass::ANCILLA,
+            }))
+        })
+    }
+
+    /// Is this an ancilla register?
+    ///
+    /// Non-ancilla registers can still contain ancilla qubits, but not the other way around.
+    pub fn is_ancilla(&self) -> bool {
+        *self.subclass() == QubitSubclass::ANCILLA
+    }
+}
+// This isn't intended for use from Rust space.
+/// Implement an ancilla register.
+#[pyclass(
+    name = "AncillaRegister",
+    module = "qiskit.circuit",
+    frozen,
+    extends=PyQuantumRegister
+)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PyAncillaRegister;
```

Comment:
If these aren't intended for use from Rust space, then I think we ought to make them `pub(crate)`, or make them private and add a PyO3 registration function to this module that gets called from `lib.rs`.

Not a hill I want to die on, though.